### PR TITLE
[TP] add support for fused QKV Sharding

### LINKED ISSUE #140069: DTensor support for fused qkv matmul
### 🚀 The feature, motivation and pitch

For transformer architecture (for example https://github.com/pytorch-labs/gpt-fast/blob/main/model.py#L195-L211) it tends to be most performant to merge the qkv matrices together. If you try to shard this concatenated tensor then the subsequent SDPA op won't be shared correctly since you need each column of q sharded with the corresponding columns of  k and v [q1,k1,v1,...], but by default the sharding will be [q1, q2, q3...] When not using DTensor this is relatively easy to get to work: https://github.com/pytorch-labs/gpt-fast/blob/main/tp.py#L73

but for DTensor the way to enable this is really unclear. is there a way to handle this type of operation with DTensor parallelization or should we just stick to normal tensor parallel support and figure out how to get it to work with our APIs?

This is currently blocking tensor parallel support in torchAO so i wanted to centralize discussion to a single location.

### Alternatives

don't use DTensor for tensor parallel

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by wanchaol**
This PR adds fused QKV sharding in the TP layer. There should be no "strided" sharding involved as fused QKV linear layer is more about combining three layers into one.

See design and discussions: https://github.com/pytorch/pytorch/issues/140069#issuecomment-2683153303

resolves https://github.com/pytorch/pytorch/issues/140069

Fixes #ISSUE_NUMBER

cc @H-Huang @awgu @kwen2501 @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: test/distributed/tensor/parallel/test_tp_style.py [hunk 1]
```diff
@@ -80,6 +80,71 @@ def test_colwise_parallel_style(self):
             )
             self.assertEqual(comm_mode.get_total_counts(), 2)
 
+    @with_comms
+    def test_colwise_parallel_fused_qkv_linear(self):
```
Comment (tianyu-l) [inline test/distributed/tensor/parallel/test_tp_style.py:84]: The test is about correctness of shapes & communications. Could you also add assertions on numerics?

File: torch/distributed/tensor/parallel/style.py [hunk 2]
```diff
@@ -118,13 +127,70 @@ def _partition_linear_fn(self, name, module, device_mesh):
...
+                        combined_sharded_weight,
+                        device_mesh,
+                        [Shard(0)],
+                        run_check=False,
+                    )
+                ),
             )
-            module.register_parameter(name, dist_param)
+            # handle the fused bias if it exists
+            if getattr(module, "bias", None) is not None:
```
Comment (tianyu-l) [inline torch/distributed/tensor/parallel/style.py:159]: The code snippets for `weight` and `bias` look identical except for the parameter name. Can we merge them into one, similar to what you did for the normal `nn.Linear` case.

File: torch/distributed/tensor/parallel/style.py [hunk 3]
```diff
@@ -54,6 +54,13 @@ class ColwiseParallel(ParallelStyle):
         output_layouts (Placement, optional):
             The DTensor layout of the output for the nn.Module, this is used to ensure the output of the nn.Module
             with the user desired layout. If not specified, the output tensor is sharded on the last dimension.
+        fused_linear_weight_splits (tuple[int, ...], optional):
```
Comment (tianyu-l) [inline torch/distributed/tensor/parallel/style.py:57]: There is one assumption made here: the fusion is along the "Column" dimension, i.e. the dimension being sharded on. Can we explicitly document this? (I understand that if fusion is not on the sharding dimension, there is no need to do custom treatment.)

**Review by tianyu-l — COMMENTED** [review]
I have one concern about this change.

I think in general DTensor `Shard` can guarantee correctness through sharding and unsharding. This PR seems taking a shortcut through TP API -- it customizes the sharding process and uses `Shard` to represent something only the user/module with `fused_linear_weight_splits` info knows.

Unless the user explicitly calls `split` with `fused_linear_weight_splits` info afterwards, there is no correctness guarantee. E.g. what if some other module calls `Shard(0) -> Replicate()`, what it gets is not the original replicated fused weights. In fact, I believe this would cause DCP resharding fail.

I think the text-book way to do this may be inventing a new `FusedShard` placement where it stores the metadata `fused_linear_weight_splits`, which handles `split` (to become `Shard(0)` given corresponding shape args) and `redistribute` rules correctly.

Or, if it is too complicated to guarantee general correctness and we only want to unblock this particular use case, I think it might be better to inherit `ColwiseParallel` to make a special `ParallelStyle` and write specific notes for it.

Please let me know what you think.

File: torch/distributed/tensor/parallel/style.py (hunk 3 above)
Comment (XilunWu) [reply torch/distributed/tensor/parallel/style.py:57]: I think "fused dimension and sharding dimension are always dim-0" is valid.

File: torch/distributed/tensor/parallel/style.py (hunk 2 above)
Comment (XilunWu) [reply torch/distributed/tensor/parallel/style.py:159]: as @tianyu-l  comment, `fused_distribute_tensor(tensor_tuple, *, mesh, dim)` can be added to increase code reuse.

File: torch/distributed/tensor/parallel/style.py [hunk 4]
```diff
@@ -118,13 +127,70 @@ def _partition_linear_fn(self, name, module, device_mesh):
...
+            combined_sharded_weight = torch.cat(sharded_weights, dim=0)
+            module.register_parameter(
+                "weight",
+                nn.Parameter(
+                    DTensor.from_local(
+                        combined_sharded_weight,
+                        device_mesh,
+                        [Shard(0)],
+                        run_check=False,
+                    )
```
Comment (XilunWu) [inline torch/distributed/tensor/parallel/style.py:155]: One thing to note is, the result DTensor may not be legal in terms of uneven sharding. Consider the case:
```
# q, k, v has shape [6, 4]
# mesh has shape [4,]
```
`q_local` after `distribute_tensor()` call will have shape `[2, 4]` on rank 0-2 and `[0, 4]` on rank 3, and `combined_sharded_weight` will have shape `[6, 4]` on rank 0-2 and `[0, 4]` on rank 3. This is different from sharding a `[18, 4]` tensor to 4 ranks (`[5, 4]` on rank 0-2 and `[3, 4]` on rank 3).

As long as we don't call `full_tensor`, we're fine.

**Review by XilunWu — COMMENTED** [review]
Thanks for supporting fused QKV weight in attention layer for Tensor Parallel!! Left some comments and one concern.

**Comment by wanchaol** [comment]
@tianyu-l @XilunWu Yes DCP reshard would not correct as it is technically not a "legal" DTensor, and I think this is the issue that is quite similar to "FlatParameter", where for some small performance reason developers might want to "fuse" multiple parameters together (more on inference), and then want to shard on it. However this patterns makes the sharding not happen on individual parameters, so it is not expressible easily with DTensor as DTensor is a per-parameter sharding primitive.

I think we don't really want to add any new Shard placement as it would significantly complicates the operators, so I did it this way that sacrifices DCP reshard correctness. Thinking more about it, I think I am leaning towards the suggestion by @tianyu-l to have a separate "FusedLinearColwiseParallel" style instead , and do not make the weights be DTensor. I'll update the PR accordingly

**Comment by hanwen-sun** [comment]
Can this pr support GQA, where the shape of q and k, v is different?

**Comment by wanchaol** [comment]
> Can this pr support GQA, where the shape of q and k, v is different?

@hanwen-sun Yes it can support GQA
